### ISSUE TITLE
anaconda-mode will open many connections

### ISSUE THREAD

**Issue by stardiviner** [proofit404/anaconda-mode#169]
for waiting anaconda server response.
When I list process in emacs, found about 15 processes.
Like this: `*127.0.0.1:port*`, `*127.0.0.1:port<1>*`, `*127.0.0.1:port<2>*` etc.

I use `company-anaconda`.
Emacs version: `24.5.1`
latest MELPA installed anaconda-mode and company-anaconda.

**Comment by proofit404** [comment]
Yes, this is a problem. I guess I can add timeout checks for killing stalled requests, but this signify more serious problem. This mean that anaconda-mode server can be unreliable for some reason.

I will return to this issue when I will have a time.

**Comment by xarthurx** [comment]
@proofit404 
Is this problem fixed?

**Comment by proofit404** [comment]
Hi @xarthurx,

I still can't reproduce this issues, for me it works perfectly. I understand that users have problems, but it requires a lot of time to figure out the root of this problem and fix it properly. I really want to help, but I have very busy schedule since I start my own business.

Regards, Artem.

**Comment by bet4it** [comment]
The reason is that, `url-http-attempt-keepalives` in `emacs` is set to `t` by default, which means that `url-retrieve` will keep a persistent connection with one target address, and follow requests with that address will be sent within this connection.But if we send two requests at the same time, it seems that `emacs` will make a new connection for the second request.
However, `service-factory` used by `anaconda-mode` can only handle one connection at one time, so if there already has a persistent connection, follow connections will be stucked.

Here is sample code to demonstrate it:

1. Save value of `anaconda-mode-server-command` to `server.py`.
2. Run `python server.py ~/.emacs.d/.cache/anaconda-mode/0.1.13 127.0.0.1 ""` and remember the port opened.
3. Open `wireshark` and filter packets with this port
4. Run following code in `emacs`. Remember to change the `anaconda-mode-port` value with actual port.
```elisp
(let* ((url-request-method "POST")
       (url-request-data "{\"jsonrpc\": \"2.0\", \"method\": \"company_complete\", \"params\": {\"source\":\"s\",\"line\":1,\"column\":1,\"path\":\"/tmp/a.py\"}, \"id\": 1}")
       (anaconda-mode-host "127.0.0.1")
       (anaconda-mode-port "12345")
       (anaconda-mode-url (format "http://%s:%s" anaconda-mode-host anaconda-mode-port)))
  (url-retrieve anaconda-mode-url #'(lambda (&rest args)) nil t)
  (url-retrieve anaconda-mode-url #'(lambda (&rest args)) nil t))
```
Then if you run `list-processes` you will see two connections, you can also see them with `ss -ntp`. And in `wireshark` you can notice that the second connection is stucked.

One workaround is using `(setq url-http-attempt-keepalives nil)` to prevent persistent connection. But in this way you will see many connections with `TIME-WAIT` state in the result of `ss -ntp`.

I think a better way to solve it is making the JSON-RPC server asynchronous.

**Comment by proofit404** [comment]
Hi @bet4it 

Thanks for your report!

I'll try it for sure.

Regards, Artem.

**Comment by Colelyman** [comment]
Just want to say that I am experiencing the same issue and @bet4it's workaround of `(setq url-http-attempt-keepalives nil)` has seemed to work for me.

Thanks!

**Comment by stardiviner** [comment]
Thanks @proofit404 , finally fixed this problem.
